File: app/src/main/java/com/example/todo/ActivityEditTodo.java
package com.example.todo;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import com.example.todo.data.TodoItem;
import com.example.todo.data.TodoItemDataUtil;
import com.example.todo.data.TodoItemTypeUtils;
import com.example.todo.databinding.ActivityEditTodoBinding;
import com.example.todo.view.TypeIndicator;

public class ActivityEditTodo extends AppCompatActivity {
    public static String TAG_TODO_ITEM = "todo";
    public static final String MODE = "mode";
    public static final int MODE_ADD = 1;
    public static final int MODE_EDIT = 2;

    private ActivityEditTodoBinding binding;
    private TodoItemDataUtil dataUtil;
    private TodoItem curr;
    public boolean isAdd = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityEditTodoBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        new Thread(() -> dataUtil = TodoItemDataUtil.getInstance(this)).start();
        Intent intent = getIntent();
        isAdd = intent.getIntExtra(MODE, MODE_ADD) == MODE_ADD;
        if (isAdd) initForAdd();
        else initForEdit(intent);
        // TODO: 周期设置 - 暂不可用
        binding.editTodoPeriodBackgroundMask.setVisibility(View.VISIBLE);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setTitle(isAdd ? R.string.add_todo : R.string.edit_todo);
        }
        binding.editTodoButtonSubmit.setOnClickListener(v -> handleSubmit());
        binding.editTodoTypeRadioGroup.setOnCheckedChangeListener((group, checkedId) -> onRadioButtonSelected(checkedId));

//        binding.editTodoPeriodStart.setOnClickListener(v -> {
//            DialogFragment fragment = new FragmentDatePicker(new DatePickerListener());
//            fragment.show(getSupportFragmentManager(), "datePicker");
//        });
//        binding.editTodoPeriodStart.setText(TimeUtils.getTodayText());
    }

    private void onRadioButtonSelected(int type) {
        int typeColor = TodoItemTypeUtils.getTypeColorByTypeId(type, this);
        binding.editTodoTypeBar.setBackground(new TypeIndicator(typeColor));
    }

    private void initForAdd() {
        onRadioButtonSelected(TodoItemTypeUtils.TYPE_NORMAL);
        binding.editTodoDeleteButton.setOnClickListener(v -> {
            Toast.makeText(ActivityEditTodo.this, "未保存", Toast.LENGTH_SHORT).show();
            finish();
        });
    }

    private void initForEdit(Intent intent) {
        curr = (TodoItem) intent.getSerializableExtra(TAG_TODO_ITEM);
        onRadioButtonSelected(curr.type);
        binding.editTodoContent.setText(curr.content);
        binding.editTodoDeleteButton.setOnClickListener(v -> {
            Toast.makeText(ActivityEditTodo.this, "已删除", Toast.LENGTH_SHORT).show();
            new Thread(() -> dataUtil.deleteTodoItem(curr)).start();
            setResult(RESULT_OK);
            finish();
        });
        if (curr.type == TodoItemTypeUtils.TYPE_PERIODIC) binding.editTodoTypeRadioGroup.setButtonsDisabled(curr.type);
        else binding.editTodoTypeRadioGroup.setButtonChecked(curr.type);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            onBackPressed();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void handleSubmit() {
        String newContent = binding.editTodoContent.getText().toString();
        int newType = binding.editTodoTypeRadioGroup.getType().typeId;
        if (newContent.trim().length() == 0) {
            Toast.makeText(this, "内容不应为空", Toast.LENGTH_SHORT).show();
            return;
        }
        if (isAdd) {
            TodoItem item = new TodoItem(newContent, newType, false);
            new Thread(() -> dataUtil.insertTodo(item)).start();
            if (newType == TodoItemTypeUtils.TYPE_PERIODIC) {
                TodoItem todayItem = new TodoItem(newContent, TodoItemTypeUtils.TYPE_NORMAL, true);
                new Thread(() -> dataUtil.insertTodo(todayItem)).start();
            }
            if (newType != TodoItemTypeUtils.TYPE_LATER) setResult(RESULT_OK);
        } else {
            if (curr.type != newType && newType == TodoItemTypeUtils.TYPE_PERIODIC) {
                TodoItem todayItem = new TodoItem(newContent, TodoItemTypeUtils.TYPE_NORMAL, true);
                new Thread(() -> dataUtil.insertTodo(todayItem)).start();
            }
            curr.content = newContent;
            curr.type = newType;
            new Thread(() -> dataUtil.updateTodoItem(curr)).start();
            setResult(RESULT_OK);
        }
        finish();
    }

    @Override
    public boolean onKeyUp(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_ENTER) {
            handleSubmit();
        }
        return false;
    }
//    class DatePickerListener implements FragmentDatePicker.DatePickerListener{
//        @Override
//        public void onDateChosen(String date) {
//            binding.editTodoPeriodStart.setText(date);
//        }
//    }
}

File: app/src/main/java/com/example/todo/ActivityEditTodo.kt
package com.example.todo

import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.example.todo.data.PeriodData
import com.example.todo.data.TodoItem
import com.example.todo.databinding.ActivityEditTodoBinding
import com.example.todo.utils.*
import com.example.todo.view.DateOrTimeSetListener
import com.example.todo.view.DatePickerFragment
import com.example.todo.view.TimePickerFragment
import com.example.todo.view.TypeIndicator

class ActivityEditTodo : AppCompatActivity() {

    companion object {
        const val TODO_OBJECT = "todo"
        const val mode = "mode"
        const val MODE_ADD = 1
        const val MODE_EDIT = 2
    }

    private lateinit var binding: ActivityEditTodoBinding
    private var curr: TodoItem? = null
    private var isAdd: Boolean = true

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityEditTodoBinding.inflate(layoutInflater)
        setContentView(binding.root)
        parseModeAndInit()
    }

    private fun onRadioButtonSelected(type: Int) {
        binding.editTodoTypeBar.background =
            TypeIndicator(TodoTypeUtil.getTypeColorByTypeId(type, this))
        if (type == TodoTypeUtil.TYPE_PERIODIC) {
            binding.editTodoPeriodBlock.visibility = View.VISIBLE
            binding.editTodoPeriodStart.text = TimeDataUtil.getTodayText()
        } else {
            binding.editTodoPeriodBlock.visibility = View.GONE
        }
    }

    private fun parseModeAndInit() {
        isAdd = intent.getIntExtra(mode, MODE_ADD) == MODE_ADD
        initView()
        if (!isAdd) initForModeEdit(intent)
    }

    private fun initView() {
        supportActionBar?.let {
            it.setDisplayHomeAsUpEnabled(true)
            it.title = resources.getText(if (isAdd) R.string.add_todo else R.string.edit_todo)
        }
        // type
        binding.editTodoTypeRadioGroup.setOnCheckedChangeListener { _, checkedId ->
            onRadioButtonSelected(checkedId)
        }
        binding.editTodoTypeRadioGroup.setButtonChecked(TodoTypeUtil.TYPE_NORMAL)
        onRadioButtonSelected(TodoTypeUtil.TYPE_NORMAL)

        // period
        binding.editTodoPeriodRadioGroup.setOnCheckedChangeListener { _, checkedId ->
            binding.editTodoPeriodUnit.text = TodoTypeUtil.getText(
                PeriodInfo.getPeriodInfo(getPeriodType(checkedId)).textId,
                this
            )
        }
        binding.editTodoPeriodDuration.setOnFocusChangeListener { _, hasFocus ->
            if (!hasFocus) getNum(
                binding.editTodoPeriodDuration.text.toString()
            )
        }
        binding.editTodoPeriodUnit.text = TodoTypeUtil.getText(PeriodInfo.PERIOD_DAY.textId, this)
        binding.editTodoPeriodStart.text = TimeDataUtil.getTodayText()
        binding.editTodoPeriodStart.setOnClickListener {
            val fragment = DatePickerFragment()
            fragment.setResultListener(object : DateOrTimeSetListener {
                override fun setResult(dateOrTimeText: String) {
                    binding.editTodoPeriodStart.text = dateOrTimeText
                    dateAddedValidCheck(TimeDataUtil.getDateToken(dateOrTimeText))
                }
            })
            fragment.show(supportFragmentManager, "datePicker")
        }
        binding.editTodoTimes.setOnFocusChangeListener { _, hasFocus ->
            if (!hasFocus) getNum(
                binding.editTodoTimes.text.toString()
            )
        }
        binding.editTodoPeriodTimesCheckbox.setOnCheckedChangeListener { _, isChecked ->
            binding.editTodoTimes.isEnabled = !isChecked
        }

        // remind
        binding.editTodoRemindTimeBlock.visibility = View.INVISIBLE
        binding.editTodoRemindTime.text = TimeDataUtil.getNowTimeText()
        binding.editTodoRemindTime.setOnClickListener {
            val fragment = TimePickerFragment()
            fragment.setResultListener(object : DateOrTimeSetListener {
                override fun setResult(dateOrTimeText: String) {
                    binding.editTodoRemindTime.text = dateOrTimeText
                }
            })
            fragment.show(supportFragmentManager, "timePicker")
        }
        binding.editTodoRemindCheckbox.setOnCheckedChangeListener { _, isChecked ->
            binding.editTodoRemindTimeBlock.visibility =
                if (isChecked) View.VISIBLE else View.INVISIBLE
        }
        binding.editTodoDeleteButton.setOnClickListener {
            Toast.makeText(this, R.string.not_saved, Toast.LENGTH_SHORT).show()
            finish()
        }
        binding.editTodoButtonSubmit.setOnClickListener { handleSubmit() }
    }

    private fun initForModeEdit(intent: Intent) {
        curr = intent.getSerializableExtra(TODO_OBJECT) as TodoItem
        curr?.let {
            binding.editTodoContent.setText(it.content)
            binding.editTodoTypeRadioGroup.setButtonChecked(it.type!!)
            onRadioButtonSelected(it.type!!)
            if (it.type == TodoTypeUtil.TYPE_PERIODIC) {
                binding.editTodoPeriodDuration.setText(it.periodValue.toString())
                binding.editTodoTimes.setText(it.periodTimes.toString())
                Toast.makeText(
                    this,
                    if (it.periodTimesLeft < 0) "无限次" else "还剩${it.periodTimesLeft}次",
                    Toast.LENGTH_SHORT
                ).show()
                binding.editTodoPeriodStart.text = TimeDataUtil.getDateText(it.dateAdded)
                binding.editTodoRemindTime.text = TimeDataUtil.getTimeText(it.remindTime)
            }
            binding.editTodoDeleteButton.setOnClickListener { _ ->
                Toast.makeText(this, R.string.deleted, Toast.LENGTH_SHORT).show()
                ThreadPoolUtil.submitTask { TodoDataUtil.deleteTodo(it, this) }
                setResult(RESULT_OK)
                finish()
            }
        }
    }

    private fun handleSubmit() {
        val content = binding.editTodoContent.text.toString()
        val type = binding.editTodoTypeRadioGroup.getType().typeId
        val periodUnit = getPeriodType(binding.editTodoPeriodRadioGroup.checkedRadioButtonId)
        val periodValue = getNum(binding.editTodoPeriodDuration.text.toString())
        val periodTimes =
            if (binding.editTodoPeriodTimesCheckbox.isChecked) -1 else getNum(binding.editTodoTimes.text.toString())
        val startTime = TimeDataUtil.getDateToken(binding.editTodoPeriodStart.text.toString())
        if (type == TodoTypeUtil.TYPE_PERIODIC && !submitValidCheck(
                content,
                periodValue,
                periodTimes,
                startTime
            )
        ) return

        val period = if (type == TodoTypeUtil.TYPE_PERIODIC) PeriodData(
            startTime,
            periodUnit,
            periodValue,
            periodTimes
        ) else null
        if (isAdd)
            TodoWrapper.getTypeWrapperById(type).handler!!.add(this, content, type, period)
        else {
            curr?.let {
                if ((periodUnit != it.periodUnit || periodValue != it.periodValue) && type == TodoTypeUtil.TYPE_PERIODIC) {
                    binding.editTodoPeriodStart.text = TimeDataUtil.getTodayText()
                    Toast.makeText(this, "周期变更后，起始日期会发生更改，请确认", Toast.LENGTH_SHORT).show()
                    return
                }
                TodoWrapper.getTypeWrapperById(type).handler!!
                    .update(this, curr!!, content, type, period)
            }
        }
        if (!isAdd || type != TodoTypeUtil.TYPE_LATER) setResult(RESULT_OK)
        finish()
    }

    private fun submitValidCheck(
        content: String,
        periodValue: Int,
        periodTimes: Int,
        startTime: Int
    ): Boolean {
        if (content.trim().isEmpty()) {
            Toast.makeText(this, R.string.content_should_not_be_empty, Toast.LENGTH_SHORT).show()
            return false
        }
        if (periodValue == 0 || periodTimes == 0) return false
        if (!dateAddedValidCheck(startTime)) return false
        return true
    }

    private fun dateAddedValidCheck(startToken: Int): Boolean {
        if (startToken < TimeDataUtil.getTodayToken()) {
            Toast.makeText(
                this@ActivityEditTodo,
                "请选择从今往后的日子！",
                Toast.LENGTH_SHORT
            ).show()
            return false
        }
        return true
    }

    private fun getPeriodType(type: Int): Int = when (type) {
        R.id.edit_todo_button_year -> TodoTypeUtil.PERIOD_YEAR
        R.id.edit_todo_button_month -> TodoTypeUtil.PERIOD_MONTH
        R.id.edit_todo_button_week -> TodoTypeUtil.PERIOD_WEEK
        else -> TodoTypeUtil.PERIOD_DAY
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            Toast.makeText(this, if (isAdd) "未保存" else "未保存更改", Toast.LENGTH_SHORT).show()
            onBackPressed()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    /**
     * return:
     *  -1  无限次
     *  0   无效
     *  other 有效
     *
     */
    private fun getNum(s: String): Int {
        if (s.isBlank()) {
            Toast.makeText(
                this@ActivityEditTodo,
                R.string.content_should_not_be_empty,
                Toast.LENGTH_SHORT
            ).show()
            return 0
        } else {
            try {
                val num = s.trim().toInt()
                when {
                    num < 1 -> Toast.makeText(
                        this@ActivityEditTodo,
                        R.string.invalid_number,
                        Toast.LENGTH_SHORT
                    ).show()
                    num > 999 -> Toast.makeText(
                        this@ActivityEditTodo,
                        R.string.number_cannot_be_too_large,
                        Toast.LENGTH_SHORT
                    ).show()
                    else -> return num
                }
                return 0
            } catch (e: Exception) {
                Toast.makeText(this@ActivityEditTodo, R.string.invalid_number, Toast.LENGTH_SHORT)
                    .show()
                return 0
            }
        }
    }
}